### PR TITLE
fix(reader): read parquet with parquet-go so a damaged file fails bounded

### LINKED ISSUE #414: A 433-byte Parquet file allocates gigabytes and never returns
Found on 966b647 while fuzzing the binary readers, alongside #413. This one is not a panic: the reader keeps allocating and never returns.

## Phenomenon

A 433-byte Parquet file holds `filesql.OpenContext` forever and grows the heap by roughly 350 MiB per second while it does. Twenty seconds in it has allocated 21 GiB with 7 GiB live, and it is still going; left alone it takes the machine's memory. A program that loads a file someone else supplied — which is what this package is for — is stopped dead by a file smaller than this issue.

Nothing in the file's own metadata looks unusual, which is what makes it hard to defend against from here: it declares 24 rows, three columns, one row group, and column chunks of about three kilobytes. The blow-up happens below that, while a page is being decoded, where this package has no visibility.

A deadline does not help. `pqarrow.FileReader.ReadTable` with a context that expires in two seconds is still running fifteen seconds later, so a caller who bounds their own work cannot bound this, and neither can this package by passing its caller's context down.

## Reproduction

```go
package repro

import (
	"context"
	"encoding/base64"
	"os"
	"path/filepath"
	"runtime"
	"testing"
	"time"

	"github.com/nao1215/filesql"
)

// 433 bytes, found by fuzzing parser.Parse over Parquet.
const slowParquet = "MDAwMBUYFQAVCBUILBUGFRAVMBUwHBhvFTAZQTkEGAYwMDAwMDAVBjAVBCMwGAIwMDcwMDAw" +
	"MDAwMDAVDCMwGAQwMDAwIzBMHDAwMBUKJTAYBTAwMDAwMBYwGRwZMCbaMBwVMBkwMDAwGRgC" +
	"MDAVABYwFtIwFtIBJjAmCBwYIDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwWAgw" +
	"MDAwMDAwMDAZIBUwFTAVMDAVMBUwFTAwMDAmjDAcFTAZMDAwMBkYBDAwMDAVABYwFrIwFrIB" +
	"JrQwJtoBHDYwFjAYBTAwMDAwGAgwMDAwMDAwMDAZIBUwFTAVMDAVMBUwFTAwMDAm3jAcFTAZ" +
	"MDAwMBkYBTAwMDAwFQAWMBbSMBbSASbYMCaMAxwYCDAwMDAwMDAwGAgwMDAwMDAwMBYwFjAY" +
	"CDAwMDAwMDAwGAgwMDAwMDAwMDAZIBUwFTAVMDAVMBUwFTAwMDAW1jAWBiYwFtYwFDAwGQwY" +
	"IjAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAZMBwwMBwwMDAwMDCRAQAAUEFS" +
	"MQ=="

func TestSlowParquetThroughTheLoader(t *testing.T) {
	data, err := base64.StdEncoding.DecodeString(slowParquet)
	if err != nil {
		t.Fatal(err)
	}
	path := filepath.Join(t.TempDir(), "slow.parquet")
	if err := os.WriteFile(path, data, 0o600); err != nil {
		t.Fatal(err)
	}

	done := make(chan error, 1)
	go func() {
		db, err := filesql.OpenContext(context.Background(), path)
		if db != nil {
			db.Close()
		}
		done <- err
	}()

	ticker := time.NewTicker(5 * time.Second)
	defer ticker.Stop()
	deadline := time.After(20 * time.Second)
	for {
		select {
		case err := <-done:
			t.Logf("finished: %v", err)
			return
		case <-ticker.C:
			var m runtime.MemStats
			runtime.ReadMemStats(&m)
			t.Logf("heap %d MiB, total allocated %d MiB", m.HeapAlloc>>20, m.TotalAlloc>>20)
		case <-deadline:
			t.Fatal("a 433 byte file has held OpenContext for 20s")
		}
	}
}
```

```
    slowloader_test.go:55:   after 5s: heap 1567 MiB, total allocated 5377 MiB
    slowloader_test.go:55:   after 10s: heap 3766 MiB, total allocated 10556 MiB
    slowloader_test.go:55:   after 15s: heap 5498 MiB, total allocated 15546 MiB
    slowloader_test.go:55:   after 20s: heap 7061 MiB, total allocated 21720 MiB
    slowloader_test.go:60: a 433 byte file has held OpenContext for 20s, heap 7061 MiB, total allocated 21720 MiB
```

The file's metadata, read with the same library:

```
    file is 433 bytes
    declared rows: 24
    row groups: 1
    columns: 3
      row group 0: rows=3 totalByteSize=3115
        column 0: values=24 totalCompressed=105 totalUncompressed=3113
        column 1: values=24 totalCompressed=89 totalUncompressed=3097
        column 2: values=24 totalCompressed=105 totalUncompressed=3113
```

## Expected behavior

A damaged file fails, in time and memory bounded by the file's own size. That is what every other reader in this package does, and what the Parquet reader does for every other kind of damage — a truncated file, a bad footer, a file of four bytes are all immediate errors.

## Cause, and why this package cannot hold the boundary

The allocation happens inside the Arrow library's page decoding, below the metadata this package can see and inside calls it cannot interrupt:

- The metadata is no help. A check that the declared rows and column-chunk sizes are plausible for the file's length passes here: 24 rows and three kilobytes of chunks in a 433-byte file are already inconsistent, but the reader does not blow up on the declared sizes — it blows up on what a page header says once decoding starts.
- A deadline is no help either, as measured above: the decode does not check the context.
- Go gives no way to stop a goroutine that is allocating, and no per-call memory budget, so wrapping the read cannot bound it.

That leaves two directions, and both are decisions for this repository rather than something to patch around:

1. Read Parquet with `github.com/parquet-go/parquet-go`, which this module already depends on (it is used in the `prep` tests to write fixtures). Whether its reader is bounded on the same input has to be measured before committing to it, and the type and NULL mapping would have to be checked against the current behavior.
2. Say in the documentation that a Parquet file from an untrusted source can exhaust memory, so a caller who accepts uploads should set `GOMEMLIMIT` and read Parquet in a process they can afford to lose.

Nothing is being reported to the Arrow project, per this repository's rule about third-party repositories. The condition for revisiting is the same either way: whether a page decode there becomes bounded by the file's size, or starts honoring the context it is given.

## Tests to add

- Whichever direction is taken, the input above belongs in the Parquet tests with a bound: the load must finish, one way or the other, within a few seconds and without allocating more than a few tens of megabytes for a 433-byte file. Written as a table it can carry the other damaged files too, so the property is "a damaged Parquet file fails quickly" rather than a single case.
- If the documentation route is taken instead, the test cannot be written, and the README's Parquet section should carry the warning where a caller deciding to accept uploads will read it.

### PR DESCRIPTION
**Pull request by nao1215**
## Summary

Parquet is now read by `github.com/parquet-go/parquet-go` instead of the Arrow reader, closing the half of #414 no check from outside could reach. The Arrow page decoder was unbounded on damaged input — the issue's 433-byte file allocated hundreds of megabytes a second for as long as it ran, ignored the context it was given, and could not be stopped from outside — and the metadata bounds check merged earlier covered only files whose metadata is visibly inconsistent, while one whose page headers alone are damaged stayed unbounded. parquet-go refuses both classes in microseconds, in time and memory bounded by the file's own size, which is what every other reader in this package does.

## Changes

- `internal/reader/parquet.go`: rewritten on parquet-go's low-level Rows API. The schema still decides each column's type; values render as before (booleans as 1/0 under the SQLite rendering, unsigned columns as their unsigned values, floats through the shared `floatText` with its infinity literal and NaN-to-NULL, DATE and TIMESTAMP as the raw counts the file stores, INT96 as epoch nanoseconds). The FLOAT16 annotation is read from the file's raw metadata because the library's schema normalization drops it; DECIMAL renders at its scale from ints or big-endian two's-complement bytes; UUID renders in canonical hex. Three recover guards turn the library's panics on inconsistent metadata into errors. The Arrow-specific machinery — `bytesReaderAt`, the metadata bounds check, the per-column table read that dodged Arrow's goroutine panics — is gone with the reader that needed it.
- `internal/reader/parquet_test.go`, `internal/reader/parquet_sqlite_test.go`: the Arrow-array unit tests are rewritten as file-level tests written with parquet-go (flat files, chunk sizes, empty tables, unsigned columns, DECIMAL, nested lists, refusal of non-Parquet bytes) plus unit tests for the new pieces (half-float widening against the IEEE 754 values, INT96 conversion, decimal spelling, FLOAT16 metadata detection). The 433-byte file from #414 is pinned as a must-fail-quickly test, and a 458-byte file found by fuzzing this branch — a MAP annotation on a node that has a physical type, which panics the library from two call paths when the node is asked its kind — is pinned beside it.
- `CHANGELOG.md`: one entry under Unreleased.

## Design Decisions

The low-level Rows API was chosen over the reflection-based GenericReader because it exposes each value's kind, definition level and column, which is what rendering needs to reproduce the previous reader's output exactly; the existing type, round-trip and damaged-file tests all pass unchanged, and sqly's suite passes against this branch. Robustness was measured rather than assumed: 300 seconds of single-worker fuzzing over the Parquet path ran 476k inputs with no hang and no panic once the two MAP-annotation guards were in, where the first run died in minutes and pinned the 458-byte find. `BenchmarkOpenParquet` (100k rows, five runs) reads 314ms/op against 313ms and allocates 219MB against 308MB — same speed, 29% less memory, more but smaller allocations (3.6M against 2.5M). The dump side still writes through Arrow; moving it too would drop the Arrow dependency from the module entirely and is left as separate work.

## Limitations

Three untested edges render differently from the Arrow reader: a TIME column renders the raw count the file stores (an integer, aligned with DATE and TIMESTAMP) rather than Arrow's marshal form, a nested or repeated field renders as a bracketed list of its leaf values, and a half-float column written by arrow-go — whose writer does not yet emit the FLOAT16 logical annotation the format defines and pyarrow writes — reads as its two raw bytes rather than as a number.

Closes #414.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved Parquet compatibility for numeric, temporal, decimal, UUID, nested, repeated, and half-precision values.
  * Malformed or corrupted files are now rejected promptly with clear errors instead of hangs or crashes.
  * Improved handling of nulls, special floating-point values, and empty tables while preserving column information.

* **Tests**
  * Added comprehensive coverage for supported data types, edge cases, malformed files, and corrupted Parquet inputs.

### CODE REVIEW

File: internal/reader/parquet.go [hunk 1]
```diff
@@ -315,132 +50,488 @@ func readParquet(src io.Reader, opts Options, emit Emit) (Result, error) {
...
+						}
+					}
+				}
+				if errors.Is(err, io.EOF) {
+					return nil
+				}
+				if err != nil {
+					return parseError(err, "failed to read table")
+				}
+			}
```
[the remark this answers is not in the document]
Comment (nao1215) [reply internal/reader/parquet.go:127]: Fixed in a33140a. A ReadRows that returns no rows, no error and no EOF into a non-empty buffer is now reported as parquet damage instead of being looped on.

File: internal/reader/parquet.go [hunk 2]
```diff
@@ -315,132 +50,488 @@ func readParquet(src io.Reader, opts Options, emit Emit) (Result, error) {
...
+					// inconsistent metadata; asking such a type its kind
+					// panics. The values still render, by their own kind.
+					col.leaf = false
+				}
+			}
+		}
+		if col.leaf {
+			col.kind = field.Type().Kind()
+			col.float16 = col.float16 || halfFloats[leafIndex]
+		}
```
[the remark this answers is not in the document]
Comment (nao1215) [reply internal/reader/parquet.go:259]: Fixed in a33140a. Both detection paths — float16Leaves over the raw metadata and the parsed-schema Float16Type case — now require a fixed length of exactly two bytes, and the length-4 and missing-length annotations are pinned as not half floats in TestFloat16Leaves.

File: internal/reader/parquet.go [hunk 3]
```diff
@@ -315,132 +50,501 @@ func readParquet(src io.Reader, opts Options, emit Emit) (Result, error) {
...
+func float16Leaves(elements []format.SchemaElement) map[int]bool {
+	out := map[int]bool{}
+	leaf := 0
+	for _, element := range elements {
+		if !element.Type.Valid {
+			continue // a group node has no physical type and holds no column
+		}
+		if _, ok := element.LogicalType.Value.(*format.Float16Type); ok &&
+			element.TypeLength.Valid && element.TypeLength.V == 2 {
+			out[leaf] = true
```
[the remark this answers is not in the document]
Comment (nao1215) [reply internal/reader/parquet.go:281]: Fixed in afa08b8. Both detection paths now require FIXED_LEN_BYTE_ARRAY alongside the two-byte length, and the BYTE_ARRAY-with-TypeLength-2 shape is pinned as not a half float in TestFloat16Leaves.
